examples/linkedlist.py

```diff
@@ -0,0 +1,65 @@
+from __future__ import print_function, absolute_import
+from numba.utils import OrderedDict
+import numpy as np
+from numba import njit
+from numba.jitclass import jitclass
+from numba import int32, deferred_type, optional
+from numba.runtime import rtsys
+
+node_type = deferred_type()
+
+spec = OrderedDict()
+spec['data'] = int32
+spec['next'] = optional(node_type)
+
+
+@jitclass(spec)
+class LinkedNode(object):
+    def __init__(self, data, next):
+        self.data = data
+        self.next = next
+
+    def prepend(self, data):
+        return LinkedNode(data, self)
+
+
+@njit
+def make_linked_node(data):
+    return LinkedNode(data, None)
+
+
+node_type.define(LinkedNode.class_type.instance_type)
```

Comment:
Another way to look at it would be to define temporary local names used for resolution of the class type:

```
spec = OrderedDict()

spec['data'] = int32
spec['next'] = optional(deferred('LinkedNode'))

@jitclass(spec)
class LinkedNode: # etc.

```

If the `jitclass` decorator is lazy enough, this could even be made to work for mutually recursive structures.